File: Python-files/04_graphics.py
"""
This file demonstrates concepts covered in the Graphics Statistical Computing:
- Basic plotting with matplotlib
- Histograms
- Scatter plots
- Box plots
- Multiple plots and subplots
- Plot customization
- Statistical visualizations
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

# Set random seed for reproducibility
np.random.seed(42)

# Basic Plotting
print("=" * 60)
print("BASIC PLOTTING")
print("=" * 60)

# Simple line plot
x = np.linspace(0, 2*np.pi, 100)
y = np.sin(x)

plt.figure(figsize=(10, 6))
plt.plot(x, y, 'b-', linewidth=2)
plt.xlabel('x')
plt.ylabel('sin(x)')
plt.title('Sine Wave')
plt.grid(True)
plt.savefig('01_sine_wave.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 01_sine_wave.png")

# Histograms
print("\n" + "=" * 60)
print("HISTOGRAMS")
print("=" * 60)

# Generate random data
data = np.random.randn(1000)

plt.figure(figsize=(10, 6))
plt.hist(data, bins=30, density=True, alpha=0.7, color='blue', edgecolor='black')
plt.xlabel('Value')
plt.ylabel('Density')
plt.title('Histogram of Normal Distribution')
plt.grid(True, alpha=0.3)

# Add normal curve
mu, sigma = data.mean(), data.std()
x_norm = np.linspace(data.min(), data.max(), 100)
plt.plot(x_norm, stats.norm.pdf(x_norm, mu, sigma), 'r-', linewidth=2, label='Normal fit')
plt.legend()

plt.savefig('02_histogram.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 02_histogram.png")


# Scatter Plots
print("\n" + "=" * 60)
print("SCATTER PLOTS")
print("=" * 60)

# Generate correlated data
n = 100
x = np.random.randn(n)
y = 2*x + np.random.randn(n)*0.5

plt.figure(figsize=(10, 6))
plt.scatter(x, y, alpha=0.6, s=50, edgecolors='black', linewidth=0.5)
plt.xlabel('X')
plt.ylabel('Y')
plt.title('Scatter Plot with Correlation')
plt.grid(True, alpha=0.3)

# Add regression line
slope, intercept = np.polyfit(x, y, 1)
x_line = np.linspace(x.min(), x.max(), 100)
plt.plot(x_line, slope*x_line + intercept, 'r--', linewidth=2, label=f'y = {slope:.2f}x + {intercept:.2f}')
plt.legend()

plt.savefig('03_scatter_plot.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 03_scatter_plot.png")


# Box Plots
print("\n" + "=" * 60)
print("BOX PLOTS")
print("=" * 60)

# Generate data for multiple groups
group1 = np.random.normal(100, 10, 100)
group2 = np.random.normal(110, 15, 100)
group3 = np.random.normal(95, 12, 100)
group4 = np.random.normal(105, 8, 100)

data_groups = [group1, group2, group3, group4]
labels = ['Group A', 'Group B', 'Group C', 'Group D']

plt.figure(figsize=(10, 6))
plt.boxplot(data_groups, labels=labels, patch_artist=True,
            boxprops=dict(facecolor='lightblue', color='blue'),
            medianprops=dict(color='red', linewidth=2),
            whiskerprops=dict(color='blue'),
            capprops=dict(color='blue'))
plt.ylabel('Values')
plt.title('Box Plot Comparison')
plt.grid(True, alpha=0.3, axis='y')
plt.savefig('04_boxplot.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 04_boxplot.png")


# Multiple Plots (Subplots)
print("\n" + "=" * 60)
print("MULTIPLE PLOTS (SUBPLOTS)")
print("=" * 60)

fig, axes = plt.subplots(2, 2, figsize=(12, 10))

# Plot 1: Line plot
x = np.linspace(0, 10, 100)
axes[0, 0].plot(x, np.sin(x), 'b-', label='sin(x)')
axes[0, 0].plot(x, np.cos(x), 'r--', label='cos(x)')
axes[0, 0].set_title('Trigonometric Functions')
axes[0, 0].set_xlabel('x')
axes[0, 0].set_ylabel('y')
axes[0, 0].legend()
axes[0, 0].grid(True, alpha=0.3)

# Plot 2: Histogram
data = np.random.exponential(2, 1000)
axes[0, 1].hist(data, bins=30, density=True, alpha=0.7, color='green', edgecolor='black')
axes[0, 1].set_title('Exponential Distribution')
axes[0, 1].set_xlabel('Value')
axes[0, 1].set_ylabel('Density')
axes[0, 1].grid(True, alpha=0.3)

# Plot 3: Scatter plot
x = np.random.randn(200)
y = np.random.randn(200)
colors = np.random.rand(200)
axes[1, 0].scatter(x, y, c=colors, cmap='viridis', alpha=0.6, s=30)
axes[1, 0].set_title('Colored Scatter Plot')
axes[1, 0].set_xlabel('X')
axes[1, 0].set_ylabel('Y')
axes[1, 0].grid(True, alpha=0.3)

# Plot 4: Bar plot
categories = ['A', 'B', 'C', 'D', 'E']
values = [23, 45, 56, 78, 32]
axes[1, 1].bar(categories, values, color='orange', edgecolor='black')
axes[1, 1].set_title('Bar Chart')
axes[1, 1].set_xlabel('Category')
axes[1, 1].set_ylabel('Value')
axes[1, 1].grid(True, alpha=0.3, axis='y')

plt.tight_layout()
plt.savefig('05_subplots.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 05_subplots.png")

# Empirical Cumulative Distribution Function (ECDF)
print("\n" + "=" * 60)
print("ECDF PLOT")
print("=" * 60)

data = np.random.randn(200)
data_sorted = np.sort(data)
ecdf = np.arange(1, len(data_sorted)+1) / len(data_sorted)

plt.figure(figsize=(10, 6))
plt.step(data_sorted, ecdf, where='post', linewidth=2)
plt.xlabel('Value')
plt.ylabel('ECDF')
plt.title('Empirical Cumulative Distribution Function')
plt.grid(True, alpha=0.3)
plt.savefig('06_ecdf.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 06_ecdf.png")

# Q-Q Plot (Quantile-Quantile Plot)
print("\n" + "=" * 60)
print("Q-Q PLOT")
print("=" * 60)

data = np.random.randn(200)

plt.figure(figsize=(10, 6))
stats.probplot(data, dist="norm", plot=plt)
plt.title('Q-Q Plot (Normal Distribution)')
plt.grid(True, alpha=0.3)
plt.savefig('07_qqplot.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 07_qqplot.png")


# Pair Plot (Scatter Matrix)
print("\n" + "=" * 60)
print("PAIR PLOT")
print("=" * 60)

# Generate multivariate data
n = 200
x1 = np.random.randn(n)
x2 = x1 + np.random.randn(n)*0.5
x3 = 2*x1 - x2 + np.random.randn(n)*0.3

data_matrix = np.column_stack([x1, x2, x3])
var_names = ['Variable 1', 'Variable 2', 'Variable 3']

fig, axes = plt.subplots(3, 3, figsize=(12, 12))

for i in range(3):
    for j in range(3):
        if i == j:
            # Diagonal: histogram
            axes[i, j].hist(data_matrix[:, i], bins=20, color='lightblue', edgecolor='black')
            axes[i, j].set_ylabel('Frequency')
        else:
            # Off-diagonal: scatter plot
            axes[i, j].scatter(data_matrix[:, j], data_matrix[:, i], alpha=0.5, s=20)
        
        if i == 2:
            axes[i, j].set_xlabel(var_names[j])
        if j == 0:
            axes[i, j].set_ylabel(var_names[i])

plt.tight_layout()
plt.savefig('08_pairplot.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 08_pairplot.png")

# Contour Plot
print("\n" + "=" * 60)
print("CONTOUR PLOT")
print("=" * 60)

# Create 2D function
x = np.linspace(-3, 3, 100)
y = np.linspace(-3, 3, 100)
X, Y = np.meshgrid(x, y)
Z = np.exp(-(X**2 + Y**2))

plt.figure(figsize=(10, 8))
contour = plt.contour(X, Y, Z, levels=10, cmap='viridis')
plt.colorbar(contour, label='Z value')
plt.xlabel('X')
plt.ylabel('Y')
plt.title('Contour Plot')
plt.grid(True, alpha=0.3)
plt.savefig('09_contour.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 09_contour.png")

# 3D Surface Plot
print("\n" + "=" * 60)
print("3D SURFACE PLOT")
print("=" * 60)

from mpl_toolkits.mplot3d import Axes3D

fig = plt.figure(figsize=(12, 8))
ax = fig.add_subplot(111, projection='3d')

# Create data
x = np.linspace(-5, 5, 50)
y = np.linspace(-5, 5, 50)
X, Y = np.meshgrid(x, y)
Z = np.sin(np.sqrt(X**2 + Y**2))

# Plot surface
surf = ax.plot_surface(X, Y, Z, cmap='coolwarm', alpha=0.8)
ax.set_xlabel('X')
ax.set_ylabel('Y')
ax.set_zlabel('Z')
ax.set_title('3D Surface Plot')
fig.colorbar(surf, shrink=0.5, aspect=5)

plt.savefig('10_surface3d.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 10_surface3d.png")

# Customized Plot Example
print("\n" + "=" * 60)
print("CUSTOMIZED PLOT")
print("=" * 60)

# Generate data
x = np.linspace(0, 10, 100)
y1 = np.exp(-x/10) * np.sin(2*x)
y2 = np.exp(-x/10) * np.cos(2*x)

plt.figure(figsize=(12, 6))

# Plot with custom styling
plt.plot(x, y1, 'b-', linewidth=2, label='Damped Sine', marker='o', 
         markevery=10, markersize=8, markerfacecolor='lightblue')
plt.plot(x, y2, 'r--', linewidth=2, label='Damped Cosine', marker='s',
         markevery=10, markersize=8, markerfacecolor='lightcoral')

# Customization
plt.xlabel('Time', fontsize=14, fontweight='bold')
plt.ylabel('Amplitude', fontsize=14, fontweight='bold')
plt.title('Damped Oscillations', fontsize=16, fontweight='bold')
plt.legend(loc='upper right', fontsize=12, framealpha=0.9)
plt.grid(True, linestyle=':', alpha=0.5)
plt.axhline(y=0, color='k', linestyle='-', linewidth=0.5)
plt.xlim([0, 10])

plt.savefig('11_customized.png', dpi=150, bbox_inches='tight')
plt.close()
print("Created: 11_customized.png")

print("\n" + "=" * 60)
print("SUMMARY")
print("=" * 60)
print("- matplotlib is the primary plotting library in Python")
print("- Supports various plot types: line, scatter, histogram, box, etc.")
print("- subplot() creates multiple plots in one figure")
print("- Extensive customization options available")
print("- Can create statistical plots: ECDF, Q-Q plots")
print("- 3D plotting capabilities with mplot3d")
print("=" * 60)
print("\nAll plots saved successfully!")

File: Julia-files/04_graphics.jl
"""
This file demonstrates concepts covered in the Graphics folder:
- Basic plotting with Plots.jl
- Histograms
- Scatter plots
- Box plots
- Multiple plots and subplots
- Plot customization
- Statistical visualizations
"""

using Plots
using Statistics
using StatsPlots
using Distributions
using Random

# Set random seed for reproducibility
Random.seed!(42)

# Basic Plotting
println("="^60)
println("BASIC PLOTTING")
println("="^60)

# Simple line plot
x = range(0, 2π, length=100)
y = sin.(x)

p1 = plot(x, y, linewidth=2, color=:blue, label="sin(x)",
          xlabel="x", ylabel="sin(x)", title="Sine Wave",
          grid=true, legend=:topright)
savefig(p1, "01_sine_wave.png")
println("Created: 01_sine_wave.png")

# Histograms
println("\n" * "="^60)
println("HISTOGRAMS")
println("="^60)

# Generate random data
data = randn(1000)

p2 = histogram(data, bins=30, normalize=:pdf, alpha=0.7, color=:blue,
               label="Data", xlabel="Value", ylabel="Density",
               title="Histogram of Normal Distribution", grid=true)

# Add normal curve
mu, sigma = mean(data), std(data)
x_norm = range(minimum(data), maximum(data), length=100)
y_norm = pdf.(Normal(mu, sigma), x_norm)
plot!(p2, x_norm, y_norm, linewidth=2, color=:red, label="Normal fit")

savefig(p2, "02_histogram.png")
println("Created: 02_histogram.png")


# Scatter Plots
println("\n" * "="^60)
println("SCATTER PLOTS")
println("="^60)

# Generate correlated data
n = 100
x = randn(n)
y = 2 .* x .+ randn(n) .* 0.5

p3 = scatter(x, y, alpha=0.6, markersize=5, color=:blue, markerstrokewidth=0.5,
             xlabel="X", ylabel="Y", title="Scatter Plot with Correlation",
             grid=true, legend=:topleft, label="Data")

# Add regression line
using GLM, DataFrames
df_temp = DataFrame(X=x, Y=y)
model = lm(@formula(Y ~ X), df_temp)
slope = coef(model)[2]
intercept = coef(model)[1]
x_line = range(minimum(x), maximum(x), length=100)
plot!(p3, x_line, intercept .+ slope .* x_line, linewidth=2, color=:red,
      linestyle=:dash, label="y = $(round(slope, digits=2))x + $(round(intercept, digits=2))")

savefig(p3, "03_scatter_plot.png")
println("Created: 03_scatter_plot.png")


# Box Plots
println("\n" * "="^60)
println("BOX PLOTS")
println("="^60)

# Generate data for multiple groups
group1 = randn(100) .* 10 .+ 100
group2 = randn(100) .* 15 .+ 110
group3 = randn(100) .* 12 .+ 95
group4 = randn(100) .* 8 .+ 105

data_groups = [group1, group2, group3, group4]
labels = ["Group A", "Group B", "Group C", "Group D"]

p4 = boxplot(labels, data_groups, ylabel="Values", title="Box Plot Comparison",
             grid=true, legend=false, fillalpha=0.75, linewidth=1.5)

savefig(p4, "04_boxplot.png")
println("Created: 04_boxplot.png")


# Multiple Plots (Subplots)
println("\n" * "="^60)
println("MULTIPLE PLOTS (SUBPLOTS)")
println("="^60)

# Plot 1: Line plot
x = range(0, 10, length=100)
p5_1 = plot(x, sin.(x), linewidth=2, color=:blue, label="sin(x)")
plot!(p5_1, x, cos.(x), linewidth=2, color=:red, linestyle=:dash, label="cos(x)")
plot!(p5_1, xlabel="x", ylabel="y", title="Trigonometric Functions",
      grid=true, legend=:topright)

# Plot 2: Histogram
data = rand(Exponential(2), 1000)
p5_2 = histogram(data, bins=30, normalize=:pdf, alpha=0.7, color=:green,
                 xlabel="Value", ylabel="Density",
                 title="Exponential Distribution", grid=true, legend=false)

# Plot 3: Scatter plot
x = randn(200)
y = randn(200)
colors = rand(200)
p5_3 = scatter(x, y, marker_z=colors, color=:viridis, alpha=0.6, markersize=3,
               xlabel="X", ylabel="Y", title="Colored Scatter Plot",
               grid=true, legend=false, colorbar=false)

# Plot 4: Bar plot
categories = ["A", "B", "C", "D", "E"]
values = [23, 45, 56, 78, 32]
p5_4 = bar(categories, values, color=:orange, xlabel="Category",
           ylabel="Value", title="Bar Chart", grid=true, legend=false)

p5 = plot(p5_1, p5_2, p5_3, p5_4, layout=(2, 2), size=(1200, 1000))
savefig(p5, "05_subplots.png")
println("Created: 05_subplots.png")

# Empirical Cumulative Distribution Function (ECDF)
println("\n" * "="^60)
println("ECDF PLOT")
println("="^60)

data = randn(200)
data_sorted = sort(data)
ecdf_values = (1:length(data_sorted)) ./ length(data_sorted)

p6 = plot(data_sorted, ecdf_values, linewidth=2, seriestype=:steppost,
          xlabel="Value", ylabel="ECDF",
          title="Empirical Cumulative Distribution Function",
          grid=true, legend=false)

savefig(p6, "06_ecdf.png")
println("Created: 06_ecdf.png")

# Q-Q Plot (Quantile-Quantile Plot)
println("\n" * "="^60)
println("Q-Q PLOT")
println("="^60)

data = randn(200)

p7 = qqplot(Normal(0, 1), data, xlabel="Theoretical Quantiles",
            ylabel="Sample Quantiles", title="Q-Q Plot (Normal Distribution)",
            grid=true, legend=false, markersize=4)

savefig(p7, "07_qqplot.png")
println("Created: 07_qqplot.png")


# Pair Plot (Scatter Matrix)
println("\n" * "="^60)
println("PAIR PLOT")
println("="^60)

# Generate multivariate data
n = 200
x1 = randn(n)
x2 = x1 .+ randn(n) .* 0.5
x3 = 2 .* x1 .- x2 .+ randn(n) .* 0.3

df = DataFrame(Variable1=x1, Variable2=x2, Variable3=x3)

p8 = @df df cornerplot([:Variable1 :Variable2 :Variable3],
                       grid=true, markersize=2, alpha=0.5)

savefig(p8, "08_pairplot.png")
println("Created: 08_pairplot.png")

# Contour Plot
println("\n" * "="^60)
println("CONTOUR PLOT")
println("="^60)

# Create 2D function
x = range(-3, 3, length=100)
y = range(-3, 3, length=100)
z = [exp(-(xi^2 + yi^2)) for yi in y, xi in x]

p9 = contour(x, y, z, levels=10, fill=true, color=:viridis,
             xlabel="X", ylabel="Y", title="Contour Plot",
             grid=true, colorbar=true)

savefig(p9, "09_contour.png")
println("Created: 09_contour.png")

# 3D Surface Plot
println("\n" * "="^60)
println("3D SURFACE PLOT")
println("="^60)

# Create data
x = range(-5, 5, length=50)
y = range(-5, 5, length=50)
z = [sin(sqrt(xi^2 + yi^2)) for yi in y, xi in x]

p10 = surface(x, y, z, color=:coolwarm, alpha=0.8,
              xlabel="X", ylabel="Y", zlabel="Z",
              title="3D Surface Plot", camera=(30, 30))

savefig(p10, "10_surface3d.png")
println("Created: 10_surface3d.png")

# Customized Plot Example
println("\n" * "="^60)
println("CUSTOMIZED PLOT")
println("="^60)

# Generate data
x = range(0, 10, length=100)
y1 = exp.(-x ./ 10) .* sin.(2 .* x)
y2 = exp.(-x ./ 10) .* cos.(2 .* x)

p11 = plot(x, y1, linewidth=2, color=:blue, label="Damped Sine",
           marker=:circle, markerevery=10, markersize=8,
           xlabel="Time", ylabel="Amplitude",
           title="Damped Oscillations", grid=true, legend=:topright)

plot!(p11, x, y2, linewidth=2, color=:red, linestyle=:dash,
      label="Damped Cosine", marker=:square, markerevery=10, markersize=8)

hline!(p11, [0], color=:black, linewidth=0.5, label="")

savefig(p11, "11_customized.png")
println("Created: 11_customized.png")

# Heatmap
println("\n" * "="^60)
println("HEATMAP")
println("="^60)

# Create correlation matrix
n = 50
data_matrix = randn(n, 4)
data_matrix[:, 2] .+= data_matrix[:, 1] .* 0.5
data_matrix[:, 3] .+= data_matrix[:, 1] .* 0.3
data_matrix[:, 4] .+= data_matrix[:, 2] .* 0.4

cor_matrix = cor(data_matrix)
labels = ["Var1", "Var2", "Var3", "Var4"]

p12 = heatmap(labels, labels, cor_matrix, color=:RdBu, aspect_ratio=1,
              title="Correlation Heatmap", colorbar=true,
              clims=(-1, 1))

savefig(p12, "12_heatmap.png")
println("Created: 12_heatmap.png")

println("\n" * "="^60)
println("SUMMARY")
println("="^60)
println("- Plots.jl is the primary plotting library in Julia")
println("- Supports various plot types: line, scatter, histogram, box, etc.")
println("- layout() creates multiple plots in one figure")
println("- Extensive customization options available")
println("- Can create statistical plots: ECDF, Q-Q plots")
println("- 3D plotting capabilities built-in")
println("="^60)
println("\nAll plots saved successfully!")
